core/src/test/scala/unit/kafka/server/FetchRequestTest.scala

```diff
@@ -169,6 +171,59 @@ class FetchRequestTest extends BaseRequestTest {
   }
 
   /**
+   * Tests that down-conversions dont leak memory. Large down conversions are triggered
+   * in the server. The client closes its connection after reading partial data when the
+   * channel is muted in the server. If buffers are not released this will result in OOM.
+   */
+  @Test
+  def testDownConversionWithConnectionFailure(): Unit = {
+    val (topicPartition, leaderId) = createTopics(numTopics = 1, numPartitions = 1).head
+
+    val producer = TestUtils.createNewProducer(TestUtils.getBrokerListStrFromServers(servers),
+      retries = 5, keySerializer = new StringSerializer, valueSerializer = new ByteArraySerializer)
+    val msgValueLen = 100 * 1000
+    try {
+      val bytes = new Array[Byte](msgValueLen)
+      for (i <- 0 to 1000) {
+        producer.send(new ProducerRecord(topicPartition.topic, topicPartition.partition, "key", bytes))
+      }
+      producer.flush()
```

Comment:
We don't need this if we are calling `close`, right?

Reply:
Since the test relies on `flush` before the subsequent fetches, I thought it is better to call it explicitly. `close` looks like cleanup, so makes it less obvious.

Reply:
`close` implies `flush`, right? It's been this way since Java OutputStream.

Reply:
I didn't explain well. `produce.close()` looks like we are closing the producer to free resources. And it is the sort of thing we may move into a teardown or later in the method if we wanted to do another send. While an explicit `flush` looks like something the test relies on. Anyway, it doesn't matter now, I have updated the code to wait for the futures.